Upload file for `customization.logo.*` settings

Desvelao opened the pull request:
# Description

This PR improves and facilitates how the user can customize the logos through the ability to upload and remove files from `Settings/Configuration`. Moreover, it extends the setting description with properties related to files.

## Changes
- `customization.logo.*` settings definition

| Key                              | Description                                                   | Values  | Extensions                  | File size |
|----------------------------------|---------------------------------------------------------------|---------|-----------------------------|-----------|
| `customization.logo.app`         | Set the custom logo for the menu and other sections           | String  | `jpeg`, `jpg`, `png`, `svg` | 1 MB      |
| `customization.logo.healthcheck` | Set the custom logo for the app's health check                | String  | `jpeg`, `jpg`, `png`, `svg` | 1 MB      |
| `customization.logo.reports`     | Set the custom logo for the PDF reports                       | String  | `jpeg`, `jpg`, `png`        | 1 MB      |
| `customization.logo.sidebar`     | Set the custom logo for the category in the navigation drawer | String. | `jpeg`, `jpg`, `png`, `svg` | 1 MB      |



- Add endpoints to manage the file upload/deletion
  - Upload: `PUT /utils/configuration/files/{key}`
    - Maximum request body is `1MB`
    - File extensions and request body size restrictions
    - File extension is detected through the file buffer. The `.jpeg` files will be saved as `.jpg`.
    - Store the file in the file system and save the setting with a relative URL to the exposed asset.
    - The `customization.logo.*` files are stored in:
      - file system: `<WAZUH_PLUGIN>/public/assets/custom/images/<key><extension>`
      - exposed URL: `<PLUGIN_PLATFORM_ADDRESS>/plugins/wazuh/assets/custom/images/<key><extension>`
  - Deletion: `DELETE /utils/configuration/files/{key}`
    - Remove the file from the file system
    - Set the setting to a empty string `""`
- Add plugin setting type: `filepicker`
- Add file picker form input component
- If some `customization.logo.*` is set, then:
  - Display the customized image in `Settings/Configuration`
  - Add a button to remove the customized image
- Create a function to convert bytes to a meaningful unit.
- Extend the setting description with:
  - Supported file extensions
  - Recommended image dimensions
  - File size restrictions

- Tests
  - Add test for the `InputForm` component of type `filepicker`
  - Add test for the new endpoints:
    - `PUT /utils/configuration/files/{key`
      - Validate the received file: max size
    - `DELETE /utils/configuration/files/{key}`
      - Validate the setting: known and unknown

Closes #4434 

**Screenshots**

- Setting configuration using a file picker
![image](https://user-images.githubusercontent.com/34042064/192512455-119c6831-cef4-44c8-a897-afcbdf4d1d94.png)
- When there is a custom image set
![image](https://user-images.githubusercontent.com/34042064/188819407-1248f70d-56c0-4b15-b93c-ccf3933b1839.png)
- Hovering the remove icon
![image](https://user-images.githubusercontent.com/34042064/188819639-d989cb38-2aa7-4d9c-889b-ae63de6c30b3.png)
- Confirmation to remove the custom image
![image](https://user-images.githubusercontent.com/34042064/188819747-8f49daec-981b-441a-af58-93a02230d2fe.png)

# Tests

## Manual tests

**Scenario**: Select a file with the suggested extension for each `customization.logo.*` setting in `Settings/Configuration`
  **When** the user drag-and-drops or selects a file in the file picker of `Settings/Configuration`
  **Then** the file picker should suggest files with the supported extensions

> This scenario should be done with all the `customization.logo.*` settings.

  
**Scenario**: Upload the configuration for a `customization.logo.*` setting in `Settings/Configuration`
  **When** the user navigates to `Settings/Configuration`, selects a valid file and clicks on the `Save` button
  **Then** the file should be saved in the expected path of the file system: `WAZUH_PLUGIN_PATH/public/assets/custom/images` whose filename is the setting key and the extension is the same as the uploaded file. (`jpeg` files will be saved as `jpg`)
  **And** the image should be displayed on the UI depending on the setting customized. Keep in mind that currently, the `customization.logo.sidebar` requires reloading the browser tab.
  **And** the configuration file should be modified to include the setting and its value is a relative URL that should be exposed.

> This scenario should be done with all the `customization.logo.*` settings.


**Scenario**: Select a file that exceeds the size limit displays a validation error in `Settings/Configuration`
  **When** the user navigates to `Settings/Configuration` and selects a file that exceeds the size limit. Above 1 MB.
  **Then** a validation error should be displayed
  
> This scenario should be done with all the `customization.logo.*` settings.


**Scenario**: Delete the customization of each `customization.logo.*` setting in `Settings/Configuration`
  **Given** the setting is customized
  **When** the user navigates to `Settings/Configuration` and clicks on the button to remove the file
  **Then** the file should be removed from the file system: `WAZUH_PLUGIN_PATH/public/assets/custom/images`
  **And** the UI should display the default image. Keep in mind that currently the `customization.logo.sidebar` setting requires reloading the browser tab
  **And** the configuration file for the setting should be cleaned. This means that its value is an empty string `""`
  
> This scenario should be done with all the `customization.logo.*` settings.

  
**Scenario**: Update the plugin configuration from `Settings/Configuration` by modifying multiple settings at the same time
  **Given** the user navigates to `Settings/Configuration` and modifies some setting that is not a `customization.logo.*` and one of `customization.logo.*` setting
  **When** the user clicks on the button to save
  **Then** the file should be saved in the expected path of the file system: `WAZUH_PLUGIN_PATH/public/assets/custom/images` whose filename is the setting key and the extension is the same as the uploaded file.  (`jpeg` files will be saved as `jpg`)
  **And** the image should be displayed depending on the setting customized. Currently, the `customization.logo.sidebar` requires reloading the browser tab
  **And** the configuration file should be modified to include the setting and its value is a relative URL that should be exposed
  **And** all the modified settings should be saved in the configuration file
  
**Scenario**: Update the plugin configuration from `Settings/Configuration` modifying multiple settings at the same time with 2 or more `customization.logo.*` settings
  **Given** the user navigates to `Settings/Configuration` and modifies some setting that is not a `customization.logo.*` and 2 or more of `customization.logo.*` settings
  **When** the user clicks on the button to save
  **Then** the files should be saved in the expected path of the file system: `WAZUH_PLUGIN_PATH/public/assets/custom/images` whose filename is the setting key and the extension is the same as the uploaded file. (`jpeg` files will be saved as `jpg`)
  **And** the images should be displayed depending on the setting customized. Keep in mind that currently, the `customization.logo.sidebar` requires reloading the browser tab.
  **And** the configuration should be modified to include the settings and its value is a relative URL that should be exposed.
  **And** all the modified settings should be saved in the configuration file
  
**Scenario**: Select a valid file for each `customization.logo.*` setting in `Settings/Configuration` with different aspect ratio
  **When** the user navigates to `Settings/Configuration` and drag-and-drops or selects a file
  **Then** the file picker should suggest the supported file extensions
  **And** renders properly the uploaded custom logo, without visual glitches or styling issues

> This scenario should be done with all the `customization.logo.*` settings.
  
## Automatic tests
- `common/services/file-size.test.ts`
- `public/components/common/form/index.test.tsx`
- `server/routes/wazuh-utils/wazuh-utils.test.ts`
- `common/plugin-settings.test.ts`
- `common/services/file-extension.test.ts`
  
**Run all PR tests**
```shell
yarn test:jest common/services/file-size.test.ts public/components/common/form/index.test.tsx server/routes/wazuh-utils/wazuh-utils.test.ts common/plugin-settings.test.ts common/services/file-extension.test.ts
```

asteriscos commented:
### Different aspect ratio example

![image](https://user-images.githubusercontent.com/9343732/190233032-7049903e-bdf1-4fa9-bcc5-6ff3f244405c.png)

asteriscos commented:
## Pending

- [ ] Clear the file picker state when the file is uploaded

![image](https://user-images.githubusercontent.com/9343732/190707102-916b4379-7fd1-40db-bb20-e4e24e31fde0.png)

Desvelao commented:
# Update

- Update the development branch with the last changes of the base branch.

Desvelao commented:
# Requirements
Yesterday, in a sync meeting, we decide to get the file extension from the buffer instead of the current logic that send the file and the extension as another field.

Desvelao commented:
# Update
- Changed as the file extension was gotten when uploading the file. Now, it gets from the file buffer.
  - Create a service to get the file extension
  - Add some tests for the service
  - Remove the previous logic and adapt to use the service

Desvelao commented:
# Update
- Update the development branch with the base branch

yenienserrano reviewed (approved):
LGTM

AlexRuiz7 commented on common/constants.ts:
```diff
@@ -342,6 +342,11 @@ export const DOCUMENTATION_WEB_BASE_URL = "https://documentation.wazuh.com";
 export const ELASTIC_NAME = 'elastic';
 
+
+// Customization
```
Default 1 MB (actual value is in bytes)

Desvelao commented:
> Would you have to change to the branch you are merging to?

Changed.

Desvelao commented on public/components/settings/configuration/components/categories/components/show-toasts.tsx:
```diff
@@ -0,0 +1,52 @@
+export const toastRequiresReloadingBrowserTab = () => {
+	getToasts().add({
+	  color: 'success',
+	  title: 'This setting requires you to reload the page to take effect.',
```
The messages were not changed, they only were moved to another file, buy I will change the message with your suggestion.

Desvelao replied:
Changed.

Desvelao commented on public/components/settings/configuration/components/categories/components/show-toasts.tsx:
```diff
@@ -0,0 +1,52 @@
+	EuiFlexItem,
+} from '@elastic/eui';
+import { PLUGIN_PLATFORM_NAME } from '../../../../../../../common/constants';
+import _ from 'lodash';
```
Removed.

Desvelao commented on public/components/settings/configuration/components/categories/components/show-toasts.tsx:
```diff
@@ -0,0 +1,52 @@
+export const toastRequiresRunningHealthcheck = () => {
+	const toast = getToasts().add({
+	  color: 'warning',
+	  title: 'You must execute the health check for the changes to take effect',
```
Changed.

Desvelao commented on public/components/settings/configuration/configuration.tsx:
```diff
@@ -29,15 +29,18 @@ import {
 import { compose } from 'redux';
-import { getSettingsDefaultList, groupSettingsByCategory, getCategorySettingByTitle } from '../../../../common/services/settings';
+import { getPluginSettingDescription, getSettingsDefaultList, groupSettingsByCategory, getCategorySettingByTitle } from '../../../../common/services/settings';
+import _ from 'lodash';
```
Removed.

Desvelao commented on public/components/settings/configuration/configuration.tsx:
```diff
@@ -29,15 +29,18 @@ import {
 import { updateSelectedSettingsSection } from '../../../redux/actions/appStateActions';
 import { withUserAuthorizationPrompt, withErrorBoundary, withReduxProvider } from '../../common/hocs';
-import { PLUGIN_PLATFORM_NAME, PLUGIN_SETTINGS, PLUGIN_SETTINGS_CATEGORIES, UI_LOGGER_LEVELS, WAZUH_ROLE_ADMINISTRATOR_NAME } from '../../../../common/constants';
+import { EpluginSettingType, PLUGIN_PLATFORM_NAME, PLUGIN_SETTINGS, PLUGIN_SETTINGS_CATEGORIES, UI_LOGGER_LEVELS, WAZUH_ROLE_ADMINISTRATOR_NAME } from '../../../../common/constants';
```
Removed.

Desvelao commented on public/components/settings/configuration/configuration.tsx:
```diff
@@ -154,8 +182,29 @@ const WzConfigurationSettingsProvider = (props) => {
+          }catch(error){
+            // TODO: There is an error related to accessing to `target` of undefined
+            // that we should review it.
+          }
```
It couldn't replicate it. Maybe some previous change doesn't cause this problem anymore.

Desvelao commented:
# Update
- Fixed a problem with an unexpected attribute for the input of type number.
- Fixed a problem when removing the configuration for `customization.logo.reports` that displayed the original logo when should not.

Desvelao commented on public/components/settings/configuration/components/categories/components/show-toasts.tsx:
```diff
@@ -0,0 +1,52 @@
+export const toastRequiresRestartingPluginPlatform = () => {
+	getToasts().add({
+	  color: 'warning',
+	  title: `You must restart ${PLUGIN_PLATFORM_NAME} for the changes to take effect`,
```
Changed.

Desvelao commented on public/components/settings/configuration/configuration.tsx:
```diff
@@ -29,15 +29,18 @@ import {
 import { WzRequest } from '../../../react-services';
 import { UIErrorLog, UIErrorSeverity, UILogLevel, UI_ERROR_SEVERITIES } from '../../../react-services/error-orchestrator/types';
 import { getErrorOrchestrator } from '../../../react-services/common-services';
 import { getToasts } from '../../../kibana-services';
```
Removed.

AlexRuiz7 commented on common/services/file-extension.ts:
```diff
@@ -0,0 +1,22 @@
+			return 'jpg'; // Also jpeg
+		case '8950':
+			return 'png';
+		case '3c3f':
```
I did some research and looks like `3c73` is also a valid file signature for SVG files.

Lots of examples [here](https://www.w3schools.com/graphics/svg_examples.asp). Copy-paste and remove the HTML tags:

```xml
<svg height="400" width="450">
<path id="lineAB" d="M 100 350 l 150 -300" stroke="red" stroke-width="3" fill="none" />
  <path id="lineBC" d="M 250 50 l 150 300" stroke="red" stroke-width="3" fill="none" />
  <path d="M 175 200 l 150 0" stroke="green" stroke-width="3" fill="none" />
  <path d="M 100 350 q 150 -300 300 0" stroke="blue" stroke-width="5" fill="none" />
  <!-- Mark relevant points -->
  <g stroke="black" stroke-width="3" fill="black">
    <circle id="pointA" cx="100" cy="350" r="3" />
    <circle id="pointB" cx="250" cy="50" r="3" />
    <circle id="pointC" cx="400" cy="350" r="3" />
  </g>
  <!-- Label the points -->
  <g font-size="30" font-family="sans-serif" fill="black" stroke="none" text-anchor="middle">
    <text x="100" y="350" dx="-30">A</text>
    <text x="250" y="50" dy="-10">B</text>
    <text x="400" y="350" dx="30">C</text>
  </g>
  Sorry, your browser does not support inline SVG.
</svg>
```

![image](https://user-images.githubusercontent.com/15186973/198055884-973bc0bb-9484-41ab-8ed4-831c85aaa4c3.png)

AlexRuiz7 reviewed (changes requested):
Errors during testing:

- Delete file: 400 Bad Request
![Screenshot from 2022-10-26 16-17-18](https://user-images.githubusercontent.com/15186973/198050935-0ee3b2ac-5e6e-46ab-ae3d-04c3f562e1aa.png)


- Upload valid SVG file: 400 Bad Request
![image](https://user-images.githubusercontent.com/15186973/198051108-5aa2fb9e-3f17-4ac5-96ef-5776b2214e78.png)

Mayons95 reviewed (commented):
TR: 
<details>
[Screencast from 26-10-22 10:34:07.webm](https://user-images.githubusercontent.com/104914131/198067322-c363fa7c-7d62-4fc6-974c-4b5dd74f9345.webm)


[Screencast from 26-10-22 10:49:12.webm](https://user-images.githubusercontent.com/104914131/198067359-6f4c38fd-87d7-46d3-8cbe-53edb79ae8b7.webm)




</details>

Desvelao commented:
# Update
- Change the layout of the filepicker input and the logo with the button to remove the configuration.

**Screenshots**
:computer: Desktop
![image](https://user-images.githubusercontent.com/34042064/198071385-8b53f951-7097-4d96-ad5f-c2d5fe9d035d.png)
![image](https://user-images.githubusercontent.com/34042064/198071460-d0078097-597b-40bd-b8a6-94c3be2ea7e2.png)

:iphone: Mobile
![image](https://user-images.githubusercontent.com/34042064/198071628-4770c818-cbe5-4dc9-a56b-e1d463f0e1c6.png)
![image](https://user-images.githubusercontent.com/34042064/198071571-43ef1a93-f8e0-4985-8deb-668b5f5077e9.png)

Mayons95 commented:
TR: :heavy_check_mark: 

<details>

[Screencast from 26-10-22 13:57:01.webm](https://user-images.githubusercontent.com/104914131/198089786-a29fb484-982c-42e0-ad0f-c82f19a8010d.webm)

[Screencast from 26-10-22 13:58:09.webm](https://user-images.githubusercontent.com/104914131/198089828-6eacda3b-6d4d-49dc-89bd-95aedecb8795.webm)


</details >

Mayons95 reviewed (dismissed):
LGTM

yenienserrano reviewed (approved):
LGTM

Mayons95 commented:
TR Comments:
We performed a script with @Desvelao and @asteriscos to test the API file upload, I left the script here, to use it, it's necessary to update the cookie value.

[script.tar.gz](https://github.com/wazuh/wazuh-kibana-app/files/9879764/script.tar.gz)

Mayons95 reviewed (approved):
TR: :heavy_check_mark: 
LGTM

Desvelao commented:
Some problems found by @AlexRuiz7 

**Environment: Kibana 7.10.2 development mode.**

## The endpoint to remove the configuration for the `customization.logo.*` setting doesn't work.
Endpoint: `DELETE /utils/configuration/files/{key}`

We were researching and saw that could be caused by the development proxy (not using the `--no-base-path` when starting the server) and accessing using the `5601` port instead of the mentioned port `5603` in the logs of the platform. We do not know the reason for the problem.

Others requests with the same HTTP verb (`DELETE`) fail too.
![image](https://user-images.githubusercontent.com/34042064/198526460-92802b1e-21da-4a85-ace5-7b33957bb882.png)

## Some custom logos are not visible when using SVG images.
This problem happens in:
- custom image preview in `Settings/Configuration`
- plugin menu

This is caused because the SVG image has not defined the `height` and `width` properties.
![image](https://user-images.githubusercontent.com/34042064/198527050-9122a7e2-d693-495a-bd81-ede4efa17977.png)

If these properties are added, the problem disappears.
![image](https://user-images.githubusercontent.com/34042064/198527168-65a99e7b-49b5-43e7-8ecb-c4835dd4e237.png)
![image](https://user-images.githubusercontent.com/34042064/198527231-30131295-18bc-499e-a401-5c408c642124.png)

@AlexRuiz7 reported that by using the same SVG for other settings as `customization.logo.healthcheck` and `customization.logo.sidebar`, the images are displayed correctly. This means that some styles could be affecting how the browser should render the image.

### Research - Custom image preview in `Settings/Configuration`

I found that if the `width: auto` CSS property is removed from the element style, the image is displayed correctly. This property is set by the `EuiImage` component when the `size` property is defined as a number https://github.com/elastic/eui/blob/v29.3.2/src/components/image/image.tsx#L140.

EDIT: I pushed a commit to solve this problem.

### Research -  Plugin menu

EDIT: I pushed a commit to solve this problem.

Desvelao commented:
In a meeting with @asteriscos , we saw a problem related to the browser cache and the `customization.logo.*` settings used in the frontend UI.

How to replicate:
1. Upload an image with some extension (`jpg`, `png`, `svg`)
2. Upload another image with the same extension as step 1. The UI display doesn't change and displays the image of step one. Cleaning the browser cache solves the problem.

We want to research a solution that does not implicate cleaning the browser cache and minimizes the users.

Desvelao commented:
> In a meeting with @asteriscos , we saw a problem related to the browser cache and the `customization.logo.*` settings used in the frontend UI.
> 
> How to replicate:
> 
> 1. Upload an image with some extension (`jpg`, `png`, `svg`)
> 2. Upload another image with the same extension as step 1. The UI display doesn't change and displays the image of step one. Cleaning the browser cache solves the problem.
> 
> We want to research a solution that does not implicate cleaning the browser cache and minimizes the users.

Solved in https://github.com/wazuh/wazuh-kibana-app/pull/4504/commits/ccd57ca7c2d00ec43eb199f452ddf5a1b42b5188.

asteriscos commented:
> > In a meeting with @asteriscos , we saw a problem related to the browser cache and the `customization.logo.*` settings used in the frontend UI.
> > How to replicate:
> > 
> > 1. Upload an image with some extension (`jpg`, `png`, `svg`)
> > 2. Upload another image with the same extension as step 1. The UI display doesn't change and displays the image of step one. Cleaning the browser cache solves the problem.
> > 
> > We want to research a solution that does not implicate cleaning the browser cache and minimizes the users.
> 
> Solved in [ccd57ca](https://github.com/wazuh/wazuh-kibana-app/commit/ccd57ca7c2d00ec43eb199f452ddf5a1b42b5188).

Now when an image is uploaded, the path saved in the configuration includes a version number to handle browser cache issues.
`customization.logo.app: "custom/images/customization.logo.app.png?v=1666948554416"`

AlexRuiz7 reviewed (approved):
- Code review: ✔️ 
- Test: ✔️

Mayons95 reviewed (approved):
LGTM 
TR: :heavy_check_mark: 
Test with the last implemented changes
